src/backend/core/api/serializers.py

```diff
@@ -226,6 +229,23 @@ def validate_file(self, file):
                 f"File type '{mime_type:s}' is not allowed. Allowed types are: {mime_types:s}"
             )
 
+        # Compress the image if it’s a valid image type
+        if mime_type and mime_type.startswith("image/"):
+            original_extension = os.path.splitext(file.name)[1]
+            image = Image.open(file)
+            image_io = BytesIO()
+            image.save(image_io, format=image.format, quality=85, optimize=True)
+            image_io.seek(0)
+            image_io.name = f"compressed{original_extension}"
+            file = image_io  # Replace the file with the compressed version
+
+            # Validate compressed file size
+            if file.getbuffer().nbytes > settings.DOCUMENT_IMAGE_MAX_SIZE:
```

Comment:
I don't think we want to reject an image just for its size as long as it is compressed.
What is missing is probably to resize the image in number of pixels and keep only what we deem necessary for the use cas of docs? We should probably use a library to do this?

Reply:
@sampaccoud In that case my Idea would be to simply add another step to limit the resolution before the compression, as "from PIL import Image" can already do this for us, there should be no need for another lib in my opinion.

What do you think of this instead?

```python3
class FileUploadSerializer(serializers.Serializer):
    """Receive file upload requests with validation, resizing, and compression."""

    file = serializers.FileField()

    def validate_file(self, file):
        """Validate, resize, and compress the file if it’s an image."""
        # Validate initial file size
        if file.size > settings.DOCUMENT_IMAGE_MAX_SIZE:
            max_size = settings.DOCUMENT_IMAGE_MAX_SIZE // (1024 * 1024)
            raise serializers.ValidationError(
                f"File size exceeds the maximum limit of {max_size:d} MB."
            )

        # Validate file type
        mime_type, _ = mimetypes.guess_type(file.name)
        if mime_type not in settings.DOCUMENT_IMAGE_ALLOWED_MIME_TYPES:
            mime_types = ", ".join(settings.DOCUMENT_IMAGE_ALLOWED_MIME_TYPES)
            raise serializers.ValidationError(
                f"File type '{mime_type:s}' is not allowed. Allowed types are: {mime_types:s}"
            )

        # Resize and compress the image if it’s a valid image type
        if mime_type and mime_type.startswith("image/"):
            original_extension = os.path.splitext(file.name)[1]
            image = Image.open(file)

            # Resize the image
            max_width, max_height = settings.DOCUMENT_IMAGE_MAX_DIMENSIONS
            image.thumbnail((max_width, max_height))

            # Compress the resized image
            image_io = BytesIO()
            image.save(image_io, format=image.format, quality=85, optimize=True)
            image_io.seek(0)
            image_io.name = f"compressed{original_extension}"
            file = image_io  # Replace the file with the resized and compressed version

            # Validate resized and compressed file size
            if file.getbuffer().nbytes > settings.DOCUMENT_IMAGE_MAX_SIZE:
                max_size = settings.DOCUMENT_IMAGE_MAX_SIZE // (1024 * 1024)
                raise serializers.ValidationError(
                    f"Compressed file size exceeds the maximum limit of {max_size:d} MB."
                )

        return file
```

I set DOCUMENT_IMAGE_MAX_DIMENSIONS to (1920, 1080). After testing, I found that the saved image size on disk didn’t exceed ~700 KB, even when the uploaded source image had a resolution of around 6000x4000 pixels. With the proposed solution, we would save the 6000x4000 source image at a resolution of 1620×1080 plus the additional compression, also keeping the dimension intact.

We could also implement a loop to compress the image until it fits the limits. But in my opinion, with the solution proposed above, we should never even come close to 10MB ...

Please let me know what you think.

Reply:
I don't think we want to reject an image because of its size. We can always resize it. Have you looked how we could use [easythumbnails](https://pypi.org/project/easy-thumbnails/) or equivalent? The original image would be deleted after a while and the thumbnailed image would remain?

Reply:
My two cents: yes it would be nice to have something on the shelf, but also, we are not doing something that complex. easy-thumbnails is useful for ... thumbnails ^^ and I'm not sure we want something that complex (because it comes with a lot of thumbnail specific enhancements (see the default processors https://github.com/SmileyChris/easy-thumbnails/blob/master/easy_thumbnails/conf.py#L234-L240) and also with a Django model field to spare time. 

PIL is that thing on the shelf ;)

Under the hood, for the resize, easy thumbnails only call the PIL `resize` method (https://github.com/SmileyChris/easy-thumbnails/blob/master/easy_thumbnails/processors.py#L252).